src/core/shutdown.c

```diff
@@ -159,6 +164,102 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+/* Read the following fields from /proc/meminfo:
+ *
+ *  NFS_Unstable
+ *  Writeback
+ *  Dirty
+ *
+ * Return true if the sum of these fields is greater than the previous
+ * value input. For all other issues, report the failure and indicate that
+ * the sync is not making progress.
+ */
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_fclose_ FILE *f = NULL;
+        char line[LINE_MAX];
+        int r = false;
+        unsigned long long val = 0;
+
+        f = fopen("/proc/meminfo", "re");
+        if (!f) {
+                return log_warning_errno(errno, "Failed to open /proc/meminfo: %m");
+        }
+
+        FOREACH_LINE(line, f, goto fail) {
+                unsigned long long ull = 0;
+
+                if (first_word(line, "NFS_Unstable:") || first_word(line, "Writeback:") || first_word(line, "Dirty:")) {
+                        if (sscanf(line, "%*s %llu %*s", &ull) != 1)
+                                goto fail;
+                } else
+                        continue;
+
+                val += ull;
```

Comment:
hmm, sounds like a strange order here... in the interest of keeping the indentation level low, i'd propose:

```c
FOREACH_LINE(…) {
        if (!first_word(line, …) && !first_word(line, …) && …)
                continue;
        if (sscanf(line, …) != 1) {
                log_error("…");
                return false;
        }

        val += …;
}
```

Reply:
Sure, implemented the above in next revision. To save trouble of adding the return, what about just replacing the log_error and return above with a "goto fail;"?

Reply:
looking at the current code there was only a single goto to "fail", so I figured there's little reason to keep the label

Reply:
Ahhh, makes sense.

Next version will remove the label and replace it with log_error_errno("...") entries for the FOREACH_LINE() macro and sscanf() errors.

Reply:
i am an idiot. i didn't see the the "goto fail" in the FOREACH_LINE parameter... sorry for the confusion. So i figure it does make some sense to have the label afte rall. That said, in the FOREACH_LINE case the error that happened is a kernel error, i.e. errno set. In the sscanf() error its quite possibly a parsing error instead, and iirc errno isn't set in that way, hence i figure it's hard to handle all that in a single label?